### ISSUE TITLE
Edit Site: Spinning up a variation on any given template

### ISSUE THREAD

**Issue by mtias** [WordPress/gutenberg#19259]
Once you are in a specific context within `edit-site` (or a context has been loaded) there should be a way to create a _variation_ for that specific context (if it's `singular`, a `singular-{post-id}`), if it's a monthly view of `archive` a `date` template and so on. This flow is decidedly more advanced, so it should be handled carefully and be easy to disable, but it should be possible.

These templates would be saved in the CPT for the site's `wp_templates` directly.

**Comment by MichaelArestad** [comment]
This is a tough one because I believe there will be more opportunities for this flow as we build out more features around templates (like a mosaic view and template part library, etc).

### Proposal i1

My first proposal uses the block toolbar to duplicate a Template. There are some caveats as I have made some assumptions by moving both the template selector and the page selector part of the template block toolbar (instead of the top toolbar). However, I believe the core of this flow could be the same regardless of where those selectors end up.

[Try the prototype](https://www.figma.com/proto/TR15iRvk7toGHnKb68QN1K/Full-Site-Editing?node-id=971%3A817&viewport=949%2C391%2C0.5&scaling=min-zoom)

#### Screenshots

<img width="1118" alt="image" src="https://user-images.githubusercontent.com/1123119/75926469-d516d400-5e27-11ea-9c80-95417c6ad3c5.png">

<img width="1126" alt="image" src="https://user-images.githubusercontent.com/1123119/75926505-e233c300-5e27-11ea-8302-694c0305f2e3.png">

#### Benefits:
* Could be an identical flow to duplicating a template partial.
* Doesn't rely on a modal
* Pretty fast

[Source figma file & frame](https://www.figma.com/file/TR15iRvk7toGHnKb68QN1K/Full-Site-Editing?node-id=971%3A817)

**Comment by johnstonphilip** [comment]
This is a good approach so far!

One thing that stands out to me in general is the issue of template hierarchy, and that it's not something average users should need to understand in order to use full-site-editing effectively and easily. 

With that in mind, I wonder if the "Save Template As" option could be a set of possible options (based on the custom post types that are currently registered). 

For example, instead of allowing people to call the templates anything they want, maybe a list like this should be available to them:

- Home (front-page)
- Single posts (single)
- List of posts (archive)
- Single Products in WooCommerce (single-product)
- List of Products from WooCommerce (archive-product)

Ideally I think this should be some kind of engaging UI that clearly indicates which templates fall under which Custom Post Types. I don't think a simple dropdown menu will be enough.

**Comment by MichaelArestad** [comment]
> With that in mind, I wonder if the "Save Template As" option could be a set of possible options (based on the custom post types that are currently registered).

@johnstonphilip I see what you're saying. We may need the user to specify what type of page it is. That's interesting. I wonder if that is necessary or if any template could be used for any page. @mtias, could you please shed some light on this?

**Comment by MichaelArestad** [comment]
After some back and forth iteration on this with @shaunandrews, here's where I landed at the end of today.

### Proposal i2
This is primarily an iteration on the Template menu. It shows the current template and useful actions one could take with that template including "Duplicate," "Rename," and "Browse templates." 

<img width="1147" alt="image" src="https://user-images.githubusercontent.com/1123119/76034447-2db7a100-5efc-11ea-8f3a-640a6d14ed04.png">

It does not include template parts. There is already two methods for selecting template parts via the Navigator and by directly selecting them. It's still possible to include them, but to me, feels unnecessary and distracts from the focus of the menu.

I also tried a variant with the added ability to change what page the user is on. I'm not entirely sold on it and it comes with its own set of issues for another... issue.

<img width="1149" alt="image" src="https://user-images.githubusercontent.com/1123119/76034434-22647580-5efc-11ea-9822-7e7d8d934b09.png">

**Comment by jasmussen** [comment]
Thanks for keeping with the high fidelity mockups! I've said it before but it deserves repeating, it makes for SUCH better conversation when you have something clear to look at.

In this case I think the duplicate/rename actions are really interesting, and would be nice to explore keeping in a section of that dropdown.

But I do find myself missing some of the pieces of #20470, including the template parts. Specifically, when the dropdown says "Home", to me it suggests that not only should "Home" also appear in the dropdown, but it should appear next to other menu items of the same DNA, i.e. other template parts. The idea being when you pick another template part, that's what becomes the label in the dropdown — after a reload of course. The additional actions can still be there if necessary, but they need to be cordoned off.

By the way the bold indicator for active item is a step up from what I have, but is there any way we can make it even more clear that it's active, without showing a checkmark like in the preview dropdown? 

Specifically, here's what I have currently, and it's bad :soblol:

<img width="305" alt="Screenshot 2020-03-06 at 09 29 03" src="https://user-images.githubusercontent.com/1204802/76066303-a7e43600-5f8d-11ea-8022-a789655eb485.png">

<img width="336" alt="Screenshot 2020-03-06 at 09 29 13" src="https://user-images.githubusercontent.com/1204802/76066308-a9adf980-5f8d-11ea-83cd-a5ea796c8163.png">

**Comment by karmatosed** [comment]
I have to note a concern I am starting to have of multiple dropdowns revealing information on the screen. I understand why we are doing this, but it's important to think about what is crucial information and what is ok to hide. We might be ok right now still, but it's a slope I'd caution against going too far down.

I understand why we are including duplication but repeating 'home' to me seems not needed. If it is needed to indication what is active then the treatment where it shows should reflect that. For example "Template: Home". 

I do prefer having this over having the template actions under 'save' which I just saw in another issue. Along with this, I would like to add a +1 to iterating on the active state.

**Comment by mtias** [comment]
It's also worth starting to look at what the settings menu would show and whether some actions for the current template belong there:

![image](https://user-images.githubusercontent.com/548849/76146291-c6cbf080-6091-11ea-8ded-17df27f1b351.png)

**Comment by mapk** [comment]
@MichaelArestad, I think the UI has changed a bit since the last round of mockups, can you provide something more recent for review?

**Comment by MichaelArestad** [comment]
@mapk Sure! Here's a minor update showing what we could do now:

<img width="290" alt="image" src="https://user-images.githubusercontent.com/1123119/87700712-04353e80-c754-11ea-9ce9-a2b44b50c138.png">

In the future, with some [enhancements to the sidebar](https://github.com/WordPress/gutenberg/issues/23939), we could potentially move template switching and some of the actions out of this menu.

**Comment by mapk** [comment]
Thanks, @MichaelArestad! I think we'll need to re-label "TOOLS" though. We use that word to describe too many other actions in the UI. Maybe we use "ACTIONS"?

**Comment by MichaelArestad** [comment]
@mapk Yeah. Maybe Actions would be better. I'm also adding "New Template":

Updated mock:

<img width="284" alt="image" src="https://user-images.githubusercontent.com/1123119/87827026-34094280-c837-11ea-99c3-f275eaced52e.png">

**Comment by mapk** [comment]
@MichaelArestad are we changing direction here to work these into the sidebar menu from https://github.com/WordPress/gutenberg/issues/23939 ?

**Comment by MichaelArestad** [comment]
@mapk eventually, yes. I'd like to see these actions exist there. Particularly the duplicate action.

**Comment by jasmussen** [comment]
Michael, your latest mock highlights the need for better systematization of where icons are located in the menu item component. When you have a moment, I'd love your thoughts on https://github.com/WordPress/gutenberg/pull/23552, which puts them on the right.

**Comment by noahtallen** [comment]
@MichaelArestad Do you think these settings would be more in line with the center-of-the-toolbar document settings explored in #20877?

**Comment by jameskoster** [comment]
I've recently been exploring how we might bring the post category edit experience to the block editor, and this issue obviously features heavily in that experience. The interesting thing about that flow in particular is that it becomes possible to edit content and templates _at the same time_. So with that context I'd like to share another option that is - I think - more inline with @MichaelArestad's i1.

Considering that this functionality is not unique to the top-level template, and will likely need to be applied to template parts as well, I think it makes sense for the following operations (and perhaps more) to be local to the selected template or template part:

* Switch instance, e.g. Switch "Header" to "Header with small logo". Or switch "Archive" to "Infinite scroll archive"
* Discard local changes to a template or template part prior to saving, without affecting any content edits
* Create a new template or template part

Here's how editing the header template part - while editing a post category - might look:

![gif](https://cldup.com/wWUv-poN9i.gif)

Note: The edit sequences (changing the background color and removing the site logo) were shortened.

Clearly, it is possible for an author to miss these actions - change a template or template part, and save their document without understanding the global reach of their changes. So I believe this interface (or something very similar) should also be surfaced during the save process. Most likely this would align with the multi-entity saving efforts.

I think this flow could also work as a part of #20477 as well - assuming it will be possible to edit templates from mosaic view.

**Comment by MichaelArestad** [comment]
@jameskoster Nice! What led you to add template part switching functionality there?

It could definitely work. It makes [renaming the template](https://github.com/WordPress/gutenberg/issues/23871#issuecomment-664495761) a little less seamless, but maybe the trade-off is worth it? I'm not sure how common template part switching would be as the context of template parts can vary. It would almost be like replacing a paragraph with an image.

**Comment by jameskoster** [comment]
> What led you to add template part switching functionality there?

It seems to me that in many cases where template parts are rendered and clickable in the editor - particularly when editing templates themselves - there should be an affordance to rotate between available template part variations. E.g. I'm creating a template for a specific post category that is initially based on the standard archive template - with this pattern it is easy to switch the base header template part with something else, or even create an entirely new header template part if I need to.

I'm not sure how seamless template part renaming needs to be? It doesn't strike me as a regular activity. It could perhaps live alongside the other actions in the popover?

<img src="https://cldup.com/nzVlJWBw17-3000x3000.png" width="350"/>

**Comment by boemedia** [comment]
> @mapk Yeah. Maybe Actions would be better. I'm also adding "New Template":
> 
> Updated mock:
> 
> <img alt="image" width="284" src="https://user-images.githubusercontent.com/1123119/87827026-34094280-c837-11ea-99c3-f275eaced52e.png">

I agree that ‘Actions’ as a label seems to reflect more what’s in there. However, I suggest renaming ‘new template’ to ‘create new’ to align the actions with the same grammar style.

**Comment by shaunandrews** [comment]
>  It makes renaming the template a little less seamless

I don't think this would have to affect renaming; You could still click the name label itself to edit, and the arrow to open the menu.

That said, I wonder if it would made sense to use the existing transform UI to list out compatible template parts?

<img width="410" alt="image" src="https://user-images.githubusercontent.com/191598/91768521-b792af00-ebab-11ea-9b25-633ad05d9b85.png">

**Comment by noahtallen** [comment]
> use the existing transform UI to list out compatible template parts

what would it mean for an existing template part to be compatible??

**Comment by jameskoster** [comment]
> That said, I wonder if it would make sense to use the existing transform UI to list out compatible template parts?

I considered the transform tool, but felt it was perhaps inappropriate for this behaviour because it does a fundamentally different job - transforming one block _type_ into another. 

It's a technicality of course - and perhaps something that is lost on the end user - but after switching a template part... it is still a template part, just a different type of template part 😆. This particular action (template part switching) felt one level removed from changing the block type as it is local to the active block. Conceptually it feels more akin to the style tool, but that isn't right either. 

That's what led me to a different pattern. I didn't want to undermine the transform tool or obfuscate its purpose. It's very possible that I overthought this though, and the transform tool would in fact be fine. No strong feelings from me on that.

> what would it mean for an existing template part to be compatible??

That is a very good question. I believe we can observe the semantic meaning generally applied to template parts in traditional theme / plugin building, and potentially use that to establish template part _types_, and consequently compatibility. Examples being things like headers, footers, content, and sidebars. Of course not all template parts will fit into these groups, and therefore not have any compatibility. I think that is fine though.

**Comment by Addison-Stavlo** [comment]
Re:
> list out compatible template parts

#24990 adds a button on the toolbar to open the already existing Template Part search/preview that is already used in the Template Part placeholder.  This could be a good starting point to iterate on in the future.  As template parts expand with *types* this search/preview component could be updated to support that, and in the case of its use for an existing template part could initialize its search state based on the current template part's *type*.

**Comment by jameskoster** [comment]
I think it's quite important that (at least for now) we only present the option to create template variations for specific contexts when we have a clear indication of the user intent for that content. This will help avoid complex UIs for selecting context when this is not apparent. 

The most reliable way to do this now would be to _only offer an option to create a specific template when a user has loaded a data scope in to the template being edited_. Example: I'm editing the Single template and have loaded the "Hello World" blog post as context. In this scenario I am able to quickly and easily spawn a new template specifically for my "Hello World" post. Here's how that might work:

![1](https://cldup.com/rvA2_G-atJ.gif)

It should be noted that we can probably generate nicer template names than "single-post-$slug" :D 

This mechanism would allow folks to spin up template variations in a variety of different situations, albeit with slightly different labels on the menu item. Here's some more examples:

### Editing the Index template and loaded the "Uncategorized" post category

<img src="https://cldup.com/08sISNEcpP-3000x3000.png" width="400" />

### Editing the Archive template and loaded the "admin" author

<img src="https://cldup.com/Vfe9eNAd32-3000x3000.png" width="400" />

### Editing the Taxonomy template and loaded the "Footwear" product category

<img src="https://cldup.com/ArvoviyC4U-3000x3000.png" width="550" />

Like the template names themselves, I'm sure we can optimise the labelling on these menu items.

---

Another use case to consider here is the idea of duplicating a template to create a new generic ($custom) template that can be applied to any singular content:

![3](https://cldup.com/D0bVJ1MaR1.gif)

Whether we implement the functionality to create both generic and specific templates initially is probably worth discussing. I imagine that either (or both) would form a solid foundation to build out more complex template variation creation in the future.

Thoughts?

**Comment by noahtallen** [comment]
Nice exploration! 

I think the "duplicate template for $entity" might be weird depending on how you switch the context. if you only switch the entity context via the preview button, it would be odd to mention the entity here (imo)

The "duplicate template for general use" is interesting, because it would basically require the template to _not_ have a slug so that it wouldn't be used anywhere. I think that is an interesting use case, but I do wonder what the "apply template to page" part of the interaction would look like, and where that ought to go. Would you be marking the template as a general template (e.g. telling it to act as the "singular" template), or overriding the template on a page?

**Comment by jameskoster** [comment]
Yeah I'm beginning to think that entity switching should [occur in the template menu](https://github.com/WordPress/gutenberg/issues/25872#issuecomment-708279824).

The general-use templates are an equivalent to the existing `$custom.php` templates. They're never automatically assigned to any content and must be explicitly chosen from the template selection dropdown in the post / page attributes meta box. The "Full width" template is a classic example that most themes include.

> Would you be marking the template as a general template (e.g. telling it to act as the "singular" template), or overriding the template on a page?

You'd be creating a new custom template, and assigning that template to the content in question. The assignment is optional because you might want to create the template now, but not use it immediately.

**Comment by jameskoster** [comment]
Re-iterating the design proposal and marking this as ready for dev :) 

When content has been loaded whilst editing a template, a "Create new template" section should appear in the template menu. 

For now, let's focus on adding a menu item there to create a template for the loaded context. IE if you loaded your "Hello World!" post in to the singular template you'd see this:

<img width="1308" alt="Screenshot 2020-11-19 at 14 59 27" src="https://user-images.githubusercontent.com/846565/99683065-d81a2900-2a77-11eb-9f45-43b4affa25ef.png">

In the future we'll likely add an option to create a new `$custom` template here as well, but we don't need to prioritise that just yet.

**Comment by jameskoster** [comment]
A supplementary (or alternative) option would be to replace the "Post" or "Page" tab in the Inspector with a "Template" tab (when editing a template, obviously). That tab could house a multitude of options, including one to "fork" the current template as described in the OP.

<img width="831" alt="Screenshot 2020-12-15 at 14 38 32" src="https://user-images.githubusercontent.com/846565/102229201-68b32000-3ee3-11eb-8c04-9c4f932593f6.png">
